Document RankTwoTensor.h

hugary1995 opened the pull request:
- Add a dummy Tensors.md page so that `RankXXXTensor`s are searchable on the MOOSE website
- Add lots of doxygen comments in RankTwoTensor.h. Most of the methods now have examples.
- Remove several enable_ifs
- ~Modify Doxyfiles for TIMPI, libmesh, framework and modules.~
- Add libMesh namespace TensorValue, TypeTensor, and TypeVector so that doxygen can generate autolinks.
  
ref idaholab/moose#21330

hugary1995 commented:
This PR will require a change to the doxygen recipe. The recipe shall look like something like this:

```sh
#!/usr/bin/env bash

# Assume I am in $MOOSE_DIR
export MOOSE_DIR=$HOME/projects/moose

# First build TIMPI doc
# This will output to $MOOSE_DIR/docs/html and also write a tag file $MOOSE_DIR/docs/timpi.tag
echo "Generating TIMPI doc"
cd $MOOSE_DIR/libmesh/contrib/timpi
doxygen $MOOSE_DIR/framework/doc/doxygen/Doxyfile_timpi > /dev/null 2>&1
mv $MOOSE_DIR/docs/html $MOOSE_DIR/docs/timpi

# Then build libmesh doc
# This will output to $MOOSE_DIR/docs/html and also write a tag file $MOOSE_DIR/docs/libmesh.tag
echo "Generating libmesh doc"
cd $MOOSE_DIR/libmesh
doxygen $MOOSE_DIR/framework/doc/doxygen/Doxyfile_libmesh > /dev/null 2>&1
mv $MOOSE_DIR/docs/html $MOOSE_DIR/docs/libmesh

# Then build framework doc
# This will output to $MOOSE_DIR/docs/html and also write a tag file $MOOSE_DIR/docs/framework.tag
echo "Generating framework doc"
cd $MOOSE_DIR/framework
doxygen $MOOSE_DIR/framework/doc/doxygen/Doxyfile > /dev/null 2>&1
mv $MOOSE_DIR/docs/html $MOOSE_DIR/docs/framework

# Finally build modules doc
# This will output to $MOOSE_DIR/docs/html
echo "Generating modules doc"
cd $MOOSE_DIR/modules
doxygen $MOOSE_DIR/modules/doc/doxygen/Doxyfile > /dev/null 2>&1
mv $MOOSE_DIR/docs/html $MOOSE_DIR/docs/modules
```

The main idea is to mimic the relative path on the host machine. I have tested the above script on my local machine, and it works. I am able to navigate to a libmesh page by clicking a link in the framework doc.

hugary1995 commented:
@bwspenc @dschwen @sapitts could any of you take a look at the doxygen comments I added? You can look at the header file directly, or if you wish you could pull this branch and generate doxygen on your own. I only looked at `RankTwoTensor.h` in this PR, and apparently there are four other classes that need care. I think this should be a good starting point for future improvements to the documentation.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1084494/) on 94d5321 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>41e7a8</th>
      <th colspan=4>#21370 94d532</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.48%</td>
      <td>83.48%</td>
      <td>+0.00%</td>
      <td>82.61%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77407</td>
      <td>77408</td>
      <td>+1</td>
      <td>57</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15318</td>
      <td>15318</td>
      <td>-</td>
      <td>12</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21370/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21370/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21370/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 82.61% is less than the suggested 90.0%



This comment will be updated on new commits.

hugary1995 committed:
Document RankTwoTensor.h

- Add a dummy Tensors.md page so that `RankXXXTensor`s are searchable on the MOOSE website
- Add lots of doxygen comments in RankTwoTensor.h. Most of the methods now have examples.
- Remove several enable_ifs
- Modify Doxyfiles for TIMPI, libmesh, framework and modules.

ref idaholab/moose#21330

sapitts reviewed (commented):
Thanks for this significant addition to the tensor documentation!

hugary1995 commented on framework/doc/content/source/utils/Tensors.md:
```diff
@@ -0,0 +1,31 @@
+- [`RankThreeTensor`/`ADRankThreeTensor`](https://mooseframework.inl.gov/docs/doxygen/moose/classRankThreeTensorTempl.html)
+- [`RankFourTensor`/`ADRankFourTensor`](https://mooseframework.inl.gov/docs/doxygen/moose/classRankFourTensorTempl.html)
+- [`SymmetricRankTwoTensor`/`ADSymmetricRankTwoTensor`](https://mooseframework.inl.gov/docs/doxygen/moose/classSymmetricRankTwoTensorTempl.html)
+- [`SymmetricRankFourTensor`/`ADSymmetricRankFourTensor`](https://mooseframework.inl.gov/docs/doxygen/moose/classSymmetricRankFourTensorTempl.html)
```
Agreed.

hugary1995 replied:
> I couldn't find this page in the Civet built documentation. Would you include a link to this page on the Tensor Mechanics module "home page" please?

Yeah, it worked on my laptop. But the website uses google search instead, which can't seem to find this page. I'll link it from the tensor mechanics homepage.

hugary1995 committed:
Address comments

hugary1995 commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * //       0 0 0 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 6`, the second order tensor is filled symmetrically using the Voigt
```
Good catch.

hugary1995 commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * //       0 0 0 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * When `input.size() == 3`, the scalar values are used to fill the in-plane components of the
```
Good catch.

dschwen commented on framework/include/utils/RankTwoTensorImplementation.h:
```diff
@@ -1033,26 +1033,27 @@ RankTwoTensorTempl<T>::initRandom(unsigned int rand_seed)
+RankTwoTensorTempl<T>::genRandomSymmTensor(T stddev, T mean)
 {
-  auto r = [&]() { return (MooseRandom::rand() + offset) * scale; };
+  auto r = [&]() { return (MooseRandom::rand() + mean) * stddev; };
```
`MooseRandom::rand()` returns a uniform distribution over [0,1), so the new parameter names are not correct.

hugary1995 replied:
Yeah you're right. 

I give up on the documentation of these two methods... I don't want to lie before we fix them.

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -76,27 +76,32 @@ void mooseSetToZero<ADRankTwoTensor>(ADRankTwoTensor & v);
+  /// @{
+
+  /**
+   * @brief Tensor dimension, i.e. number of rows/columns of the second order tensor
```
IMO "dimension" is the right word. https://dmol.pub/math/tensors-and-shapes.html#:~:text=Always%20use%20the%20word%20rank,tensor%20combines%20all%20of%20these.

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -111,37 +116,124 @@ class RankTwoTensorTempl : public TensorValue<T>
+   *
+   * Deprecated in favor of MooseRandom::seed().
+   */
+  static void initRandom(unsigned int);
```
Sidebar: this whole global seed business is not great. We should be passing around references to "random number providers". A change in seed here might affect many other systems that don't get the seed reset explicitly, and the seed set here might not even affect future random tensor draws, if some other system overrides the global seed.

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -111,37 +116,124 @@ class RankTwoTensorTempl : public TensorValue<T>
+   *
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~{.cpp}
+   * RankTwoTensor A(RankTwoTensor::initNone);
+   * // A = [ 0 0 0
+   * //       0 0 0
+   * //       0 0 0 ]
```
`initNone` should leave the data _uninitialized_. It is to be used whenever you are setting the tensor entries to custom values rightaway.

hugary1995 replied:
Fixed

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -173,85 +281,265 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * RealVectorValue row2(4, 5, 6);
+   * RealVectorValue row3(7, 8, 9);
+   * auto A = RankTwoTensor::initializeSymmetric(row1, row2, row3);
+   * // A = [ 1 3 5
+   * //       3 5 7
+   * //       5 7 9 ]
```
That doesn't look right. all 5s on the antidiagonal? 2 4 6 and 8 don't appear? They should appear as average terms.

dschwen replied:
Oh, I see, you took those as values not as placeholder indices....

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * //       5 4 3 ]
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   */
+  void fillFromScalarVariable(const VariableValue & scalar_variable);
```
Sidebar: can we combine `fillFromInputVector` and `fillFromScalarVariable` ? (I thought I had already done that). If we make the argument type a template type and take advantage of _duck typing_. I.e. the passed in type must have `.size()` and `operator[]`. I thing Both `std::vector<T>` and `VariableValue` do. This would also enable use of `std::initializer_list` and `std::array` for example...

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
-   * where the set of indices {n,o,p,q} is a permutation of the indices {i,j,k,l}
+   * Return the general tensor product of this second order tensor and another second order tensor
+   * defined as \f$ C_{ijkl} = A_{\mathcal{M}(n)\mathcal{M}(o)} B_{\mathcal{M}(p)\mathcal{M}(q)} \f$
+   * where the multiplication order is defined by the index map \f$ \mathcal{M}: \{n,o,p,q\} \to
+   * \{i,j,k,l\} \f$. The index map is specified using the template parameters. See examples below
```
This drops the requirement that the mnop indices need to be a permutation ijkl (formerly described as a bijective map). This should appear in the docs in some form. If the requirement is not fulfilled  the result might be unexpected to the user.

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * usingTensorIndices(i, j, k, l);
+   * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+   *
+   * The outer product of A and B is defined as \f$ A_{ij} B_{kl} \f$, hence the template
+   * specialization should be `times<i, j, k, l>`
```
template instantiation. A specialization is something different.

dschwen commented on framework/include/utils/RankTwoTensor.h:
```diff
@@ -262,149 +550,703 @@ class RankTwoTensorTempl : public TensorValue<T>
+   * @brief Sin(3*Lode_angle)
+   *
    * If secondInvariant() <= r0 then return r0_value
    * This is to gaurd against precision-loss errors.
```
```suggestion
   * This is to guard against precision-loss errors.
```

dschwen reviewed (dismissed):
:+1:

hugary1995 commented:
Conflicts resolved.

dschwen commented:
> Conflicts resolved.

```
error: no member named 'mixedProductIkJl' in 'RankTwoTensorTempl<double>'
```

not successfully it appears :-(

hugary1995 committed:
Resolve conflicts -- second try

dschwen commented:
Merging this should fix the unit test issue
